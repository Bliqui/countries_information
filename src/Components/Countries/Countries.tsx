import { SimpleGrid, Spinner, Text, Flex } from "@chakra-ui/react";
import { useContext, useEffect } from "react";
import { CountriesContext } from "../../context/context";
import { useFetch } from "../../hooks/useFetch";
import { DataType } from "../DataType";
import { CountryPreview } from "./CountryPreview/CountryPreview";

export const Countries = ({
  searchResult,
  selectedOption,
}: {
  searchResult: string;
  selectedOption: string;
}) => {
  const {
    data: receivedCountries,
    isLoading,
    error,
  } = useFetch<DataType[]>("https://restcountries.com/v3.1/all");

  const { state, dispatch } = useContext(CountriesContext);

  // const isRequestNeeded = () => {
  //   if (state.countries.length === receivedCountries?.length) {
  //     return state.countries;
  //   }

  // };

  useEffect(() => {
    dispatch({
      type: "ADD_COUNTRIES",
      payload: receivedCountries as [],
    });
  }, [receivedCountries]);

  let sortedCountries =
    state.countries &&
    state.countries.filter((country) => {
      if (selectedOption === "") {
        return true;
      }
      return country.region === selectedOption;
    });

  let filteredCountries =
    sortedCountries &&
    sortedCountries.filter(({ name }) => {
      return name.common.toLowerCase().includes(searchResult.toLowerCase());
    });

  return (
    <Flex flexDir="column" mt="50px">
      {error && <Text>{error}</Text>}
      {isLoading ? (
        <Spinner data-testid="spinner" mt="20%" alignSelf="center" size="lg" />
      ) : (
        <SimpleGrid
          alignItems="center"
          rowGap="80px"
          columnGap={{ sm: "20px", md: "40px", lg: "60px" }}
          minChildWidth="320px"
        >
          {filteredCountries?.length > 0 ? (
            filteredCountries.map(
              ({ name, flags, population, region, capital }) => {
                return (
                  <CountryPreview
                    key={name.common}
                    src={flags.png}
                    title={name.common}
                    population={population}
                    region={region}
                    capital={capital?.[0] ? capital?.[0] : ""}
                  />
                );
              }
            )
          ) : (
            <Text color="red.500" textAlign="center">
              No countries found
            </Text>
          )}
        </SimpleGrid>
      )}
    </Flex>
  );
};
