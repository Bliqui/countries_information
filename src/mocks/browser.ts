import { rest } from "msw";
import { setupServer } from "msw/node";

type AllCountriesDataType = {
  name: string;
  population: number;
  flag: string;
  region: string;
  capital: string;
};

type CountryPageDataType = {
  country: {
    name: string;
    capital: string;
    nativeName: string;
    population: number;
    region: string;
    subRegion: string;
    topLevelDomain: string[];
    currencies: { code: string; name: string; symbol: string }[];
    languages: {
      name: string;
      nativeName: string;
    }[];
    flag: string;
    borders: string[];
    latlng: [number, number];
  };
};

export const server = setupServer(
  rest.get<AllCountriesDataType>(
    "https://restcountries.com/v3.1/all",
    (req, res, ctx) => {
      return res(
        ctx.json([
          {
            name: { common: "Poland" },
            population: 40000000,
            flags: { png: "url" },
            region: "Europe",
            capital: ["Warsaw"],
          },
          {
            name: { common: "Russia" },
            population: 60000000,
            flags: { png: "url" },
            region: "Europe",
            capital: ["Moscow"],
          },
        ])
      );
    }
  ),
  rest.get<AllCountriesDataType>(
    `https://newsapi.org/v2/top-headlines?country=RUS`,
    (req, res, ctx) => {
      return res(
        ctx.json({
          articles: [
            {
              author: "Petya",
              title: "Shock content",
              description: "Shock content happened",
              url: "some url",
              publishedAt: "2022-08-29T08:00:28Z",
            },
            {
              author: "Vasya",
              title: "extra Shock content",
              description: "Shock content happened twice",
              url: "some url2",
              publishedAt: "2022-09-29T08:00:28Z",
            },
          ],
        })
      );
    }
  )
);
// Describe the requests to mock.
