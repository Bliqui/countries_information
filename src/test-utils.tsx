import * as React from "react";
import { render, RenderOptions } from "@testing-library/react";
import { ChakraProvider, theme } from "@chakra-ui/react";
import { AppProvider } from "./context/context";
import { BrowserRouter } from "react-router-dom";

const AllProviders = ({ children }: { children?: React.ReactNode }) => (
  <AppProvider>
    <BrowserRouter>
      <ChakraProvider theme={theme}>{children}</ChakraProvider>
    </BrowserRouter>
  </AppProvider>
);

const customRender = (ui: React.ReactElement, options?: RenderOptions) =>
  render(ui, { wrapper: AllProviders, ...options });

export { customRender };
