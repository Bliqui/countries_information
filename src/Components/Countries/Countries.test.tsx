import {
  screen,
  render,
  waitForElementToBeRemoved,
} from "@testing-library/react";
import { Countries } from "./Countries";
import { server } from "../../mocks/browser";
import { customRender } from "../../test-utils";

beforeAll(() => {
  // Establish requests interception layer before all tests.
  server.listen();
});

afterEach(() => {
  server.resetHandlers();
});

afterAll(() => {
  // Clean up after all tests are done, preventing this
  // interception layer from affecting irrelevant tests.
  server.close();
});

describe("Countries component", () => {
  it("should show loader, hide it and show list of countries after", async () => {
    customRender(<Countries searchResult="Poland" selectedOption="Europe" />);
    expect(await screen.findByTestId("spinner")).toBeInTheDocument();

    await waitForElementToBeRemoved(() => screen.queryByTestId("spinner"));

    expect(await screen.findByText(/Poland/i)).toBeInTheDocument();
    expect(await screen.findByText(/Warsaw/i)).toBeInTheDocument();
    expect(await screen.findByText(/Europe/i)).toBeInTheDocument();
  });

  it('shows "No countries found" if invalid value passed trough the search input', async () => {
    render(<Countries searchResult="ABCD" selectedOption="Europe" />);

    expect(await screen.findByTestId("spinner")).toBeInTheDocument();
    await waitForElementToBeRemoved(() => screen.queryByTestId("spinner"));

    expect(screen.getByText("No countries found")).toBeInTheDocument();
  });

  it("matches snapshot", async () => {
    const output = customRender(
      <Countries searchResult="Poland" selectedOption="Europe" />
    );
    expect(await screen.findByTestId("spinner")).toBeInTheDocument();
    await waitForElementToBeRemoved(() => screen.queryByTestId("spinner"));

    expect(await screen.findByText(/Poland/)).toBeInTheDocument();

    expect(output).toMatchSnapshot();
  });
});
