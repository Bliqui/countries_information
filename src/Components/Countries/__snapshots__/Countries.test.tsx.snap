// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Countries component matches snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body
    class="chakra-ui-light"
  >
    <div>
      <div
        class="css-1f8h9z4"
      >
        <div
          class="css-19ybp5c"
        >
          <div
            class="chakra-stack css-1hcswde"
            data-testid="countryCard"
          >
            <div
              class="css-wm8dfs"
            >
              <img
                class="chakra-image css-z2bnja"
                src="url"
              />
            </div>
            <div
              class="css-w9il2w"
            >
              <h2
                class="chakra-heading css-sawm00"
              >
                Poland
              </h2>
              <div
                class="css-0"
              >
                <div
                  class="css-0"
                >
                  <p
                    class="chakra-text css-hzw7ps"
                  >
                    Population:
                  </p>
                  40 000 000
                </div>
                <div
                  class="css-0"
                >
                  <p
                    class="chakra-text css-hzw7ps"
                  >
                    Region:
                  </p>
                  Europe
                </div>
                <div
                  class="css-0"
                >
                  <p
                    class="chakra-text css-hzw7ps"
                  >
                    Capital:
                  </p>
                  Warsaw
                </div>
              </div>
            </div>
            <a
              class="chakra-link css-36y4gk"
              href="/country/poland"
            >
              Details
            </a>
          </div>
        </div>
      </div>
      <span
        hidden=""
        id="__chakra_env"
      />
    </div>
    <div
      class="chakra-portal"
    >
      <ul
        aria-live="polite"
        id="chakra-toast-manager-top"
        role="region"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column; margin: 0px auto;"
      />
      <ul
        aria-live="polite"
        id="chakra-toast-manager-top-left"
        role="region"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column;"
      />
      <ul
        aria-live="polite"
        id="chakra-toast-manager-top-right"
        role="region"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column;"
      />
      <ul
        aria-live="polite"
        id="chakra-toast-manager-bottom-left"
        role="region"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column;"
      />
      <ul
        aria-live="polite"
        id="chakra-toast-manager-bottom"
        role="region"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column; margin: 0px auto;"
      />
      <ul
        aria-live="polite"
        id="chakra-toast-manager-bottom-right"
        role="region"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column;"
      />
    </div>
  </body>,
  "container": <div>
    <div
      class="css-1f8h9z4"
    >
      <div
        class="css-19ybp5c"
      >
        <div
          class="chakra-stack css-1hcswde"
          data-testid="countryCard"
        >
          <div
            class="css-wm8dfs"
          >
            <img
              class="chakra-image css-z2bnja"
              src="url"
            />
          </div>
          <div
            class="css-w9il2w"
          >
            <h2
              class="chakra-heading css-sawm00"
            >
              Poland
            </h2>
            <div
              class="css-0"
            >
              <div
                class="css-0"
              >
                <p
                  class="chakra-text css-hzw7ps"
                >
                  Population:
                </p>
                40 000 000
              </div>
              <div
                class="css-0"
              >
                <p
                  class="chakra-text css-hzw7ps"
                >
                  Region:
                </p>
                Europe
              </div>
              <div
                class="css-0"
              >
                <p
                  class="chakra-text css-hzw7ps"
                >
                  Capital:
                </p>
                Warsaw
              </div>
            </div>
          </div>
          <a
            class="chakra-link css-36y4gk"
            href="/country/poland"
          >
            Details
          </a>
        </div>
      </div>
    </div>
    <span
      hidden=""
      id="__chakra_env"
    />
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
